### ISSUE TITLE
[Plugin] Azure Devops (aka TFS)

### ISSUE THREAD

**Issue by Grovespaz** [spotify/backstage#598]
## Summary

Azure Devops is Microsoft's enterprise source control, issue tracker, ci/cd, etc. all-in-one solution. There is both a cloud-hosted as well as an on-premise version, which don't differ much.

## Project website (if applicable)

https://azure.microsoft.com/nl-nl/services/devops/
https://docs.microsoft.com/en-us/rest/api/azure/devops/?view=azure-devops-rest-5.1

## Context

Once there is more of a structure and/or reference for how a CI/CD plugin should behave (ie. an implementation or two exist) I'd be happy to create a plugin for Azure DevOps.

Different teams use different parts of Azure DevOps, for example we only use source control (which has two varieties, Git and the older TFVC) and build & deploy, while teams at other organizations might also use the issue tracking Azure DevOps provides (we use Jira). We'd probably need to have this be configurable.

**Comment by freben** [comment]
Thanks for making an issue about this!

This is the kind of tool that already has a mature and substantial web interface, so it becomes an interesting discussion about how much of the functionality to expose or even replicate in Backstage. I haven't used it myself so I'll defer mostly to others to decide where the convenience / effort balance lies.

For example in terms of the CI/CD functionality, it sounds like one could at least start adding:

- Linkage out to Azure DevOps to the relevant specific pages for an item (when the inventory system is in place)
- Indicators / cards showing status of ongoing builds
- Indicators / warning callouts showing abnormal states such as broken master or flaky builds
- Overview / summary list of recent builds (initially linking out to Azure?)
- More?

**Comment by javaniecampbell** [comment]
@stefanalund this can be assigned to me I am currently working on it as we discussed in discord some time ago I forgot to add the plugin feature did the planning and found the tools that would be needed my day job got me little busy but i dedicating the weekends to backstage dev and exploration

**Comment by stefanalund** [comment]
Thanks @javaniecampbell — looking forward to seeing the result 🚀

**Comment by enricomeneghel** [comment]
Hi @javaniecampbell — any outcome to share since now? could I help?

**Comment by javaniecampbell** [comment]
@enricomeneghel sorry for the late response currently on vacation will be on returning end of this week to continue development on the plugin will get in touch when I return with the current progress I have made so far and we can collaborate then..

**Comment by raosumanth** [comment]
Hi @javaniecampbell , Is there any progress on this plugin. I was also interested in knowing more on this. Thank you

**Comment by freben** [comment]
I unassigned this ticket for now. @raosumanth @enricomeneghel or others - let us know if this is something you would like to start working on. @javaniecampbell - let us or the others know if there is partial work to make a draft PR of or similar, to base further work on

**Comment by TimmyGilissen** [comment]
Is there an update on this issue?

**Comment by freben** [comment]
Nobody has picked it up yet as far as I am aware. Let us know if you are in a position to contribute!

**Comment by bruno197** [comment]
Hey guys! Where could I see the plugin's source code to check if I can help?

**Comment by benjdlambert** [comment]
Hey @bruno197!

So there isn't a plugin for the Azure CI/CD yet, this is what this ticket is about.

However, we do have some examples of CI/CD plugins and examples of how to talk to AzureDevOps in the following locations if you want to have a crack at building a plugin:

Here's some prior art of talking to Azure:
https://github.com/backstage/backstage/blob/master/plugins/scaffolder-backend/src/scaffolder/stages/publish/azure.ts

https://github.com/backstage/backstage/blob/443416b875e550fb87a29740af4a834217d99cd7/packages/backend-common/src/reading/AzureUrlReader.ts

And here's some examples of CI/CD plugins:
https://github.com/backstage/backstage/tree/master/plugins/github-actions

https://github.com/backstage/backstage/tree/master/plugins/circle-ci

I can assign this to you if you feel that you could take this on? 🏅 :tada:

**Comment by ppejovic** [comment]
Just a suggestion but the issue title is a bit misleading. It seems what is being focused on here is the CI/CD so this would be better "branded" as an Azure Pipelines plugin. There could conceivably be separate plugins for other Azure DevOps features areas; Azure Boards, etc.

**Comment by benjdlambert** [comment]
@ppejovic yep - that sounds about right. I'm not too familiar with the Azure offering, but what you're saying about mutliple plugins for different feature sets sounds correct.

**Comment by chagretes** [comment]
I strongly agree with @ppejovic. I was looking for a way to integrate with [Azure DevOps repositories](https://azure.microsoft.com/en-us/services/devops/repos/), for instance.

**Comment by adamdmharvey** [comment]
Based on @chagretes 's comment, made me realize buried in this issue is potentially multiple asks because "Azure DevOps" is a very very broad topic:

* [ ] Integration support for Azure DevOps (that is, using it as an SCM provider)
* [ ] CI/CD support to list Azure DevOps Pipelines in a CI/CD switcher/tab/widget
* [ ] Possible Artifact lister for I might offer that what you lean to there MIGHT be different than this plugin record.

Maybe this should be spun out into individual issues with a set boundary for each?  Could make it much easier to walk a milestone path towards greater Azure DevOps suite support.  (e.g., TechDocs already supports Azure)  I am not an active Azure user, so I don't quite know all the mechanics, but maybe someone could take a stab and we clarify the main descriptions of each??

**Comment by benjdlambert** [comment]
I'm pretty sure that we support integration with reading and scaffolding to Azure dev ops, however the other parts like CI/CD is not built yet.

**Comment by t3hprofit** [comment]
I've been working on some tooling against Azure DevOps's APIs and I believe the components for working with yaml Pipelines via API is not complete enough to really integrate with this at the moment.

The azure-devops-go-api library hasn't been updated since October. It looks like it's autogenerated through some tooling, though the tooling or parameters for go-swagger/swagger-codegen aren't published. As it's Microsoft's "official" go API for interacting with AZDO API's, it would make sense to use it, but it doesn't appear to be maintained very frequently.

For fun I've made some changes to a local copy of the library to try and get the Create Pipeline calls working, but so far I haven't had any luck. I'd hope to be able to scaffold pipelines in AZDO as part of whatever plugin gets created. Also, if anyone decides to pick this up, let me know as I'm interested in contributing.

**Comment by t3hprofit** [comment]
> but so far I haven't had any luck

just kidding. but their error messages aren't very insightful. Got it working, but in a totally unsupported manner

**Comment by awanlin** [comment]
@freben I'd like to take this on. I'll be starting small, just a table with the latest builds. We use nearly the full suite of tooling provided by Azure DevOps - Repos, Boards, Pipelines (Builds and Releases), and Artifacts - where I work. I know we'd eventually like to see those covered in this plugin

With that in mind, I could use some help with some design things:

**One or more plugins**: Would it best best to just have one plugin that way you could use the existing Integration PAT (Personal Access Token) for Azure DevOps or sperate plugins each using their own PAT? The API calls would be different for each so I don't think you really save much having this all in one vs. broken out. Using separate PATs means you can scope them nicely but then you have a bunch to maintains and update potentially.

**TeamProject/RepoName Annotation**: One of the sort of gotchas with the Azure DevOps API is that you would need the Team Project name as well as the Repo name to make most of the calls due to the design of the URL. There's a roughly 50/50 chance that you either have just one Team Project or more than one Team Project. For use we have more than one so we have used the following pattern for our annotation: `dev.azure.com/project-slug: TeamProject/RepoName`. Does that make sense to continue to use for this?

Future roadmap ideas:

- Releases, this is the CD side of CI/CD in the Azure DevOps world
- Pull Requests, would be nice to see this by entity in the catalog but also by team
- Artifacts, this is where your packages live - NuGet, NPM, Maven, etc., would be great to see these details for Library components
- Multi-stage pipelines, this is the newest YAML based method for CI/CD in Azure DevOps but is a little on the complex side so I think this will need to come later

**Comment by awanlin** [comment]
I've just submitted a PR for the Azure DevOps backend plugin. This is an MVP that will pull out the top n builds given a repo name. Once that's approved and merged in I'll get working on the frontend plugin to display this.

https://github.com/backstage/backstage/pull/7192

**Comment by awanlin** [comment]
Just a quick update the backend plugin was approved and merged in a few days ago. I'll be starting the frontend plugin tomorrow

**Comment by awanlin** [comment]
Another update: I've just submitted the PR for the frontend portion of the Azure DevOps plugin - #7484. This will show the top n builds for a given repo. Once that's approved and merged the next item on my list will be showing Pull Requests

**Comment by awanlin** [comment]
Review of the front end PR has gone really well and we've made several improvements from that. In the mean time I've just submitted PR #7532 that will add a route for getting Pull Request data to be used by the frontend plugin

**Comment by awanlin** [comment]
Thanks @jvilimek!
